src/Raven.Server/Documents/Sharding/BucketStatsHolder.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+using System.Collections.Generic;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Voron;
+using Voron.Data.Tables;
+using Voron.Impl;
+
+namespace Raven.Server.Documents.Sharding
+{
+    internal unsafe class BucketStatsHolder
+    {
+        private readonly ShardedDocumentsStorage _storage;
+        private readonly Dictionary<int, Documents.BucketStats> _values;
+        private readonly Dictionary<int, ChangeVector> _mergedChangeVectors;
+        private DocumentsOperationContext _ctx;
+
+        public BucketStatsHolder(ShardedDocumentsStorage storage)
+        {
+            _storage = storage;
+            _values = new Dictionary<int, Documents.BucketStats>();
+            _mergedChangeVectors = new Dictionary<int, ChangeVector>();
+        }
+
+        public void UpdateBucket(int bucket, long nowTicks, long sizeChange, long numOfDocsChanged)
+        {
+            _values.TryGetValue(bucket, out var bucketStats);
```

Comment:
BucketStats is a struct? 
In that case, you are doing a lot of copying back and forth, no?
You can get a ref to the value directly and skip that. See: https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.getvaluereforadddefault?view=net-7.0

Reply:
fixed, thanks